File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/WeakRefList.java
package com.example.alexeyglushkov.taskmanager.task;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Iterator;

/**
 * Created by alexeyglushkov on 23.08.15.
 */
public class WeakRefList<T> extends ArrayList<WeakReference<T>> {

    private static final long serialVersionUID = 4962762847646156417L;

    @Override
    public Iterator<WeakReference<T>> iterator() {
        clean();
        return super.iterator();
    }

    public void clean() {
        for (int i = this.size(); --i >= 0;) {
            if (get(i).get() == null) {
                remove(i);
            }
        }
    }

    @Override
    public boolean contains(Object object) {
        if (object instanceof WeakReference) {
            return super.contains(object);
        }

        for (WeakReference<T> ref : this) {
            if (ref.get() == object) {
                return true;
            }
        }

        return false;
    }

    @Override
    public boolean remove(Object object) {
        if (object instanceof WeakReference) {
            return super.remove(object);
        }

        int i = 0;
        for (WeakReference<T> ref : this) {
            if (ref.get() == object) {
                remove(i);
                return true;
            }

            ++i;
        }

        return false;
    }
}

File: taskmanager/src/main/java/com/example/alexeyglushkov/taskmanager/task/WeakRefList.kt
package com.example.alexeyglushkov.taskmanager.task

import java.lang.ref.WeakReference
import java.util.ArrayList

/**
 * Created by alexeyglushkov on 23.08.15.
 */
class WeakRefList<T> : ArrayList<WeakReference<T>>() {

    override fun iterator(): Iterator<WeakReference<T>> {
        clean()
        return super.iterator()
    }

    fun clean() {
        var i = this.size
        while (--i >= 0) {
            if (get(i).get() == null) {
                removeAt(i)
            }
        }
    }

    override operator fun contains(`object`: Any?): Boolean {
        if (`object` is WeakReference<*>) {
            return super.contains(`object`)
        }

        for (ref in this) {
            if (ref.get() === `object`) {
                return true
            }
        }

        return false
    }

    override fun remove(`object`: Any?): Boolean {
        if (`object` is WeakReference<*>) {
            return super.remove(`object`)
        }

        var i = 0
        for (ref in this) {
            if (ref.get() === `object`) {
                removeAt(i)
                return true
            }

            ++i
        }

        return false
    }

    companion object {

        private val serialVersionUID = 4962762847646156417L
    }
}
